File: 10_numeryczny_gradient_descent/from_c++/gradient_descent.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <iomanip>
#include <functional>

// ===== Objective function =====
double f(const std::vector<double>& vec) {
    double sum_sq = 0.0;
    double sum_sin = 0.0;
    for (double x : vec) {
        sum_sq += x * x;
        sum_sin += std::sin(3.0 * x);
    }
    return sum_sq + 0.5 * sum_sin;
}

// ===== Numerical gradient (central difference) =====
std::vector<double> numerical_gradient(
    const std::function<double(const std::vector<double>&)>& func,
    const std::vector<double>& vec,
    double h = 1e-5
) {
    std::vector<double> grad(vec.size(), 0.0);
    for (size_t i = 0; i < vec.size(); ++i) {
        std::vector<double> vec_forward = vec;
        std::vector<double> vec_backward = vec;
        vec_forward[i] += h;
        vec_backward[i] -= h;
        grad[i] = (func(vec_forward) - func(vec_backward)) / (2.0 * h);
    }
    return grad;
}

// ===== Gradient descent with adaptive learning rate =====
std::pair<std::vector<double>, std::vector<std::vector<double>>> gradient_descent(
    const std::function<double(const std::vector<double>&)>& func,
    std::vector<double> start_vec,
    double init_lr = 0.1,
    int max_iter = 1000,
    double tolerance = 1e-6,
    int log_interval = 10
) {
    std::vector<double> vec = start_vec;
    double lr = init_lr;
    std::vector<std::vector<double>> history;
    history.push_back(vec);

    for (int iteration = 0; iteration < max_iter; ++iteration) {
        std::vector<double> grad = numerical_gradient(func, vec);
        std::vector<double> new_vec(vec.size());
        for (size_t i = 0; i < vec.size(); ++i) {
            new_vec[i] = vec[i] - lr * grad[i];
        }

        // If the step increases the function value -> decrease learning rate
        if (func(new_vec) > func(vec)) {
            lr *= 0.5;
            continue;
        }

        // If the step improves the function -> slightly increase learning rate
        lr *= 1.05;
        history.push_back(new_vec);

        // Log progress every `log_interval` iterations
        if (iteration % log_interval == 0) {
            std::cout << "Iter " << iteration
                << ": f(x) = " << std::fixed << std::setprecision(6) << func(vec)
                << ", lr = " << std::setprecision(5) << lr
                << ", x = [";
            for (size_t i = 0; i < vec.size(); ++i) {
                std::cout << std::setprecision(5) << vec[i];
                if (i + 1 < vec.size()) std::cout << ", ";
            }
            std::cout << "]\n";
        }

        // Stop criteria
        double move_norm = 0.0, grad_norm = 0.0;
        for (size_t i = 0; i < vec.size(); ++i) {
            move_norm += (new_vec[i] - vec[i]) * (new_vec[i] - vec[i]);
            grad_norm += grad[i] * grad[i];
        }
        move_norm = std::sqrt(move_norm);
        grad_norm = std::sqrt(grad_norm);

        if (move_norm < tolerance || grad_norm < tolerance) {
            std::cout << "Converged at iteration " << iteration << "\n";
            break;
        }

        vec = new_vec;
    }

    return { vec, history };
}

int main() {
    // Set random number generator (equivalent to np.random.seed(0))
    std::mt19937 rng(0);
    std::uniform_real_distribution<double> dist(-3.0, 3.0);

    // Start point in 5D
    std::vector<double> start_point(5);
    for (double& x : start_point) {
        x = dist(rng);
    }

    auto [minimum_vec, path] = gradient_descent(f, start_point, 0.05);

    std::cout << "\nFinal result:\n";
    std::cout << "Minimum found at: [";
    for (size_t i = 0; i < minimum_vec.size(); ++i) {
        std::cout << std::setprecision(6) << minimum_vec[i];
        if (i + 1 < minimum_vec.size()) std::cout << ", ";
    }
    std::cout << "]\n";
    std::cout << "Function value: " << std::setprecision(6) << f(minimum_vec) << "\n";

    return 0;
}

File: 10_numeryczny_gradient_descent/from_c++/Claude/GradientDescent.java
// --- All-in-One Spring Boot Application (src/main/java/com/example/demo/GradientDescent.java) ---
package org.example;

import java.util.*;
import java.util.function.Function;

public class GradientDescent {

    // ===== Objective function =====
    public static double f(double[] vec) {
        double sumSq = 0.0;
        double sumSin = 0.0;
        for (double x : vec) {
            sumSq += x * x;
            sumSin += Math.sin(3.0 * x);
        }
        return sumSq + 0.5 * sumSin;
    }

    // ===== Numerical gradient (central difference) =====
    public static double[] numericalGradient(Function<double[], Double> func,
                                             double[] vec, double h) {
        double[] grad = new double[vec.length];

        for (int i = 0; i < vec.length; i++) {
            double[] vecForward = vec.clone();
            double[] vecBackward = vec.clone();
            vecForward[i] += h;
            vecBackward[i] -= h;
            grad[i] = (func.apply(vecForward) - func.apply(vecBackward)) / (2.0 * h);
        }

        return grad;
    }

    // Overloaded method with default h value
    public static double[] numericalGradient(Function<double[], Double> func, double[] vec) {
        return numericalGradient(func, vec, 1e-5);
    }

    // ===== Result class to hold optimization results =====
    public static class OptimizationResult {
        public final double[] minimumVec;
        public final List<double[]> history;

        public OptimizationResult(double[] minimumVec, List<double[]> history) {
            this.minimumVec = minimumVec;
            this.history = history;
        }
    }

    // ===== Gradient descent with adaptive learning rate =====
    public static OptimizationResult gradientDescent(Function<double[], Double> func,
                                                     double[] startVec,
                                                     double initLr,
                                                     int maxIter,
                                                     double tolerance,
                                                     int logInterval) {
        double[] vec = startVec.clone();
        double lr = initLr;
        List<double[]> history = new ArrayList<>();
        history.add(vec.clone());

        for (int iteration = 0; iteration < maxIter; iteration++) {
            double[] grad = numericalGradient(func, vec);
            double[] newVec = new double[vec.length];

            // Calculate new vector: newVec = vec - lr * grad
            for (int i = 0; i < vec.length; i++) {
                newVec[i] = vec[i] - lr * grad[i];
            }

            // If the step increases the function value -> decrease learning rate
            if (func.apply(newVec) > func.apply(vec)) {
                lr *= 0.5;
                continue;
            }

            // If the step improves the function -> slightly increase learning rate
            lr *= 1.05;
            history.add(newVec.clone());

            // Log progress every logInterval iterations
            if (iteration % logInterval == 0) {
                System.out.printf("Iter %d: f(x) = %.6f, lr = %.5f, x = %s%n",
                        iteration, func.apply(vec), lr, Arrays.toString(vec));
            }

            // Stop criteria
            double moveNorm = 0.0;
            double gradNorm = 0.0;

            for (int i = 0; i < vec.length; i++) {
                double diff = newVec[i] - vec[i];
                moveNorm += diff * diff;
                gradNorm += grad[i] * grad[i];
            }

            moveNorm = Math.sqrt(moveNorm);
            gradNorm = Math.sqrt(gradNorm);

            if (moveNorm < tolerance || gradNorm < tolerance) {
                System.out.println("Converged at iteration " + iteration);
                break;
            }

            vec = newVec.clone();
        }

        return new OptimizationResult(vec, history);
    }

    // Overloaded method with default parameters
    public static OptimizationResult gradientDescent(Function<double[], Double> func,
                                                     double[] startVec,
                                                     double initLr) {
        return gradientDescent(func, startVec, initLr, 1000, 1e-6, 10);
    }

    public static void main(String[] args) {
        // Set random number generator (equivalent to C++ mt19937(0))
        Random rng = new Random(0);

        // Start point in 5D
        double[] startPoint = new double[5];
        for (int i = 0; i < startPoint.length; i++) {
            // Generate random values between -3.0 and 3.0
            startPoint[i] = -3.0 + rng.nextDouble() * 6.0;
        }

        // Create function reference
        Function<double[], Double> objectiveFunction = GradientDescent::f;

        OptimizationResult result = gradientDescent(objectiveFunction, startPoint, 0.05);

        System.out.println("\nFinal result:");
        System.out.print("Minimum found at: [");
        for (int i = 0; i < result.minimumVec.length; i++) {
            System.out.printf("%.6f", result.minimumVec[i]);
            if (i + 1 < result.minimumVec.length) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
        System.out.printf("Function value: %.6f%n", f(result.minimumVec));
    }
}